File: Textbook_Exercises/MATLAB/Chapter_2/Exercise_2_13/two_d_grad_wrapper_hw.m
function two_d_grad_wrapper_hw()
% two_d_grad_wrapper.m is a toy wrapper to illustrate the path
% taken by gradient descent depending on the learning rate (alpha) chosen.
% Here alpha is kept fixed and chosen by the use. The corresponding
% gradient steps, evaluated at the objective, are then plotted.  The plotted points on
% the objective turn from green to red as the algorithm converges (or
% reaches a maximum iteration count, preset to 50).
%
% (nonconvex) function here is
% g(w) = -cos(2*pi*w'*w) + w'*w
%
% This file is associated with the book
% "Machine Learning Refined", Cambridge University Press, 2016.
% by Jeremy Watt, Reza Borhani, and Aggelos Katsaggelos.


%%% runs everything %%%
run_all()


%%%%%%%%%%%% subfunctions %%%%%%%%%%%%
%%% performs gradient descent steps %%%%
    function [w,in,out] = gradient_descent(alpha,w)

    % initializations
    grad_stop = 10^-5;
    max_its = 50;
    iter = 1;
    grad = 1;
    in = [w];
    out = [-cos(2*pi*w'*w) + 2*w'*w];
    % main loop
    while norm(grad) > grad_stop && iter <= max_its

        % take gradient step
% ----> grad = 
        w = w - alpha*grad;

        % update containers
        in = [in, w];
        out = [out, -cos(2*pi*w'*w) + 2*w'*w];

        % update stopers
        iter = iter + 1;
    end
end

function run_all()
    % dials for the toy
    alpha = 10^-2;     % step length/learning rate (for gradient descent). Preset to alpha = 10^-3

    for j = 1:2
        x0 = [-.7;0];         % initial point (for gradient descent)
        if j == 2
            x0 = [.85;.85];
            alpha = 3*10^-3;
        end
        %%% perform gradient descent %%%
        [x,in,out] = gradient_descent(alpha,x0);

        %%% plot function with grad descent objective evaluations %%%
        hold on
        plot_it_all(in,out)
    end
end

%%% plots everything %%%
function plot_it_all(in,out)
    % print function
    [A,b] = make_fun();

    % print steps on surface
    plot_steps(in,out,3)
    set(gcf,'color','w');
end

%%% plots everything %%%
function [A,b] = make_fun()
    range = 1.15;                     % range over which to view surfaces
    [a1,a2] = meshgrid(-range:0.04:range);
    a1 = reshape(a1,numel(a1),1);
    a2 = reshape(a2,numel(a2),1);
    A = [a1, a2];
    A = (A.*A)*ones(2,1);
    b = -cos(2*pi*A) + 2*A;
    r = sqrt(size(b,1));
    a1 = reshape(a1,r,r);
    a2 = reshape(a2,r,r);
    b = reshape(b,r,r);
    h = surf(a1,a2,b)
    az = 35;
    el = 60;
    view(az, el);
    shading interp


    xlabel('w_1','Fontsize',18,'FontName','cmmi9')
    ylabel('w_2','Fontsize',18,'FontName','cmmi9')
    zlabel('g','Fontsize',18,'FontName','cmmi9')
    set(get(gca,'ZLabel'),'Rotation',0)
    set(gca,'FontSize',12);
    box on
    colormap gray
end

% plot descent steps on function surface
function plot_steps(in,out,dim)
    s = (1/length(out):1/length(out):1)';
    colorspec = [s.^(1),flipud(s) ,zeros(length(out),1)];
    width = (1 + s)*5;
    if dim == 2
        for i = 1:length(out)
            hold on
            plot(in(1,i),in(2,i),'o','Color',colorspec(i,:),'MarkerFaceColor',colorspec(i,:),'MarkerSize',width(i));
        end
    else % dim == 3
        for i = 1:length(out)
            hold on
            plot3(in(1,i),in(2,i),out(i),'o','Color',colorspec(i,:),'MarkerFaceColor',colorspec(i,:),'MarkerSize',width(i));
        end
    end
end

end

File: mlrefined_exercises/ed_1/Python2/Chapter_2/Exercise_2_13/two_d_grad_wrapper_hw.py
# two_d_grad_wrapper_hw is a toy wrapper to illustrate the path
 # taken by gradient descent depending on the learning rate (alpha) chosen.
 # Here alpha is kept fixed and chosen by the use. The corresponding
 # gradient steps, evaluated at the objective, are then plotted.  The plotted points on
 # the objective turn from green to red as the algorithm converges (or
 # reaches a maximum iteration count, preset to 50).
 # (nonconvex) function here is
 #
 # g(w) = -cos(2*pi*w'*w) + 2*w'*w

 # This file is associated with the book
 # "Machine Learning Refined", Cambridge University Press, 2016.
 # by Jeremy Watt, Reza Borhani, and Aggelos Katsaggelos.

from pylab import *
from mpl_toolkits.mplot3d import Axes3D

###### ML Algorithm functions ######
def gradient_descent(w0,alpha):
    w = w0
    g_path = []
    w_path = []
    w_path.append(w)
    g_path.append(-cos(2*pi*dot(w.T,w)) + 2*dot(w.T,w))

    # start gradient descent loop
    grad = 1
    iter = 1
    max_its = 50
    while linalg.norm(grad) > 10**(-5) and iter <= max_its:
        # take gradient step
# --->  grad =
        w = w - alpha*grad

        # update path containers
        w_path.append(w)
        g_path.append(-cos(2*pi*dot(w.T,w)) + 2*dot(w.T,w))
        iter+= 1
    g_path = asarray(g_path)
    g_path.shape = (iter,1)
    w_path = asarray(w_path)
    w_path.shape = (iter,2)

    # show final average gradient norm for sanity check
    s = dot(grad.T,grad)/2
    s = 'The final average norm of the gradient = ' + str(float(s))
    print(s)


    # # for use in testing if algorithm minimizing/converging properly
    # plot(asarray(obj_path))
    # show()

    return (w_path,g_path)

###### plotting functions #######
def make_function():
    global fig,ax1

    # prepare the function for plotting
    r = linspace(-1.15,1.15,300)
    s,t = meshgrid(r,r)
    s = reshape(s,(size(s),1))
    t = reshape(t,(size(t),1))
    h = concatenate((s,t),1)
    h = dot(h*h,ones((2,1)))
    b = -cos(2*pi*h) + 2*h
    s = reshape(s,(sqrt(size(s)),sqrt(size(s))))
    t = reshape(t,(sqrt(size(t)),sqrt(size(t))))
    b = reshape(b,(sqrt(size(b)),sqrt(size(b))))

    # plot the function
    fig = plt.figure(facecolor = 'white')
    ax1 = fig.add_subplot(111, projection='3d')
    ax1.plot_surface(s,t,b,cmap = 'Greys',antialiased=False) # optinal surface-smoothing args rstride=1, cstride=1,linewidth=0
    ax1.azim = 115
    ax1.elev = 70

    # pretty the figure up
    ax1.xaxis.set_rotate_label(False)
    ax1.yaxis.set_rotate_label(False)
    ax1.zaxis.set_rotate_label(False)
    ax1.get_xaxis().set_ticks([-1,1])
    ax1.get_yaxis().set_ticks([-1,1])
    ax1.set_xlabel('$w_0$   ',fontsize=20,rotation = 0,linespacing = 10)
    ax1.set_ylabel('$w_1$',fontsize=20,rotation = 0,labelpad = 50)
    ax1.set_zlabel('   $g(\mathbf{w})$',fontsize=20,rotation = 0,labelpad = 20)

def plot_steps(w_path,g_path):
    # colors for points
    ax1.plot(w_path[:,0],w_path[:,1],g_path[:,0],color = [1,0,1],linewidth = 5)   # add a little to output path so its visible on top of the surface plot
    ax1.plot(w_path[-8:-1,0],w_path[-8:-1,1],g_path[-8:-1,0],color = [1,0,0],linewidth = 5)   # add a little to output path so its visible on top of the surface plot


def main():
    make_function()                             # plot objective function

    # plot first run on surface
    alpha = 10**-2
    w0 = array([-0.7,0])
    w0.shape = (2,1)
    w_path,g_path = gradient_descent(w0,alpha)    # perform gradient descent
    plot_steps(w_path,g_path)

    # plot second run on surface
    w0 = array([0.8,-0.8])
    w0.shape = (2,1)
    w_path,g_path = gradient_descent(w0,alpha)    # perform gradient descent
    plot_steps(w_path,g_path)
    show()
main()
